Merkletree Worker

timoxley opened the pull request:
### WIP PR for visibility/discussion.

Goal is to move the CPU intensive merkletree calculations, specifically `getContents` & `getPath`, off the server thread so they don't block.

----

Current state is a transitional step of converting the existing merkletree operations, along with their callers, to be async. This should then make it easy to swap the implementation out to one which runs asynchronously in a separate worker.

----

#### Implementation Thoughts/Discussion

We can't magically do more than one slow calculation per node process, but by moving the calculation off the server process at least the server process will be free to handle new requests. So the calculations are still a bottleneck but we could add some parallelism by spawning multiple processes. Ideally, we'd use a single worker per tree calculation, limited to a small number so as to not overwhelm the system with CPU intensive processes. This would be simplest in terms of memory management as memory is freed when process ends.

An alternative that adds concurrency would be to split the work across multiple event loop ticks. This will slow down the tasks overall, as it adds some overhead and delay, but this may be negligible and we could even limit how much time it will block before yielding to the event loop. This would allow us to only use a single worker process, which simplifies the implementation. Perhaps doing this in the existing server process could be a good step 2, it will likely still add heavy congestion and affecting the server's performance, but at least the server will have the opportunity to respond to other tasks and it avoids a lot of complications introduced by adding workers.
 
Also need to decide the best way to have `getPath` work using workers – the complication being that it requires the tree structure and where does it get this from? Does it keep a cache? is it sent via unix socket/IPC? A command-line argument?

Sending the tree to a worker process over IPC and processing it as an argument to `getPath` is probably the simplest implementation, but the serialization/deserialization  overhead & memory copies will likely result in poor performance with large trees.

If we were on Node 12, it's possible we could make use of the shared memory feature of [worker threads](https://nodejs.org/api/worker_threads.html#worker_threads_worker_threads). This may require adjusting the `buildMerkleTree` algorithm to work with views over a single Buffer. Using Node 12 there may also be something smart that can be done for both the worker thread and worker process cases using [advanced serialization](https://nodejs.org/dist/latest-v12.x/docs/api/child_process.html#child_process_advanced_serialization). 

An alternative is to have a persistent merkletree worker process which maintains some cache of trees such that when the `getPath` request comes in, it can use a local cached version, and worst case just recalculate the tree if no cached version exists. This option would need to either spawn its own workers or split the algorithm work over multiple ticks otherwise this service will also have the same problem of being totally unresponsive while trees/paths are being calculated. Being a long-lived process, there's also more memory management concerns.

Perhaps there's a more efficient `getPath` algorithm which could be performed within the server process. Also possible the slowness I've seen in `getPath` is only for unrealistic, pathological cases in the tests. 

Also a non-js reimplementation of the merkletree functions may (or may not) significantly improve the walltime performance, though it doesn't magically step around the issues with passing around large trees or concurrency. Rust or C++ may be a good choice as they would be compatible with any of the above solutions via native extension or WASM.

Possible some combination of the above will delay the need to do a clustered version of the data union server.

jtakalai commented on src/merkletree/index.js:
```diff
@@ -0,0 +1,134 @@
+        return new Promise((resolve, reject) => {
+            let result
+            if (this.proc) {
+                this.proc.kill()
```
what does send function do? It kills processes, forks new workers?

timoxley replied:
Added docs and tweaked it a bit to be clearer. It's mainly about attaching the appropriate listeners and resolving the promise at the right time.

jtakalai commented on src/merkletree/index.js:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    async buildTree(treeContents) {
+        const result = await this.sendAction({
```
is `sendAction` re-entry protection? Like, everyone who calls buildTree later would get the result from same worker? `send` does generate a new promise every time, but I guess it's a promise that resolves when worker exits?

timoxley replied:
> is sendAction re-entry protection?

Not really, mainly just a wrapper around `ChildProcess#send` that defines an "Action" as the interface, and attaches the appropriate event listeners and error handling.

> I guess it's a promise that resolves when worker exits?

Correct

jtakalai commented on src/merkletree/index.js:
```diff
@@ -0,0 +1,134 @@
+ * Ensures multiple calls to `fn` with the same first argument
+ * will only execute `fn` once while `fn` resolves.
+ */
+
```
when using jsdoc, please document args too: `@param {Function} fn`

timoxley replied:
Done

jtakalai commented on src/merkletree/index.js:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+module.exports = class MerkleTreeRPCWrapper extends MerkleTree {
```
if buildTree had re-entry protection, why is limitDuplicateAsyncWork needed?

timoxley replied:
`limitDuplicateAsyncWork` is the main reentry protection, other stuff like that `proc.kill` was just sanity check, no longer needed so I'm pulling that out.

timoxley commented:
> How does the performance look? 

Walltime is quite a bit slower, I initially thought it may be due to process startup time but I profiled that and it's <200ms. 

The worker uses ~65% less CPU in the main process.

> Did offloading calculation to branch solve HTTP responsiveness? 

It does, the server is able to tick much more frequently.

> Also, how does router answer currently? Does it await for tree calculations? src/routers/communities.js I guess still has plasma.getLatestBlock() that is sync. I'm wondering if that doesn't need to change, too.

Only place the tree seems to be used is in `/:communityAddress/members/:address`, and its call to  `plasma.getProofAt(address, withdrawableBlock.blockNumber)` was already async.

timoxley commented:
Updated/hacked the install to work with node 12 by forcing all packages depending on `scrypt.js` to use version `0.3.0` which should gracefully fall back to a JS implementation, and also forcing all versions of `scrypt` to be "optional" i.e. don't fail everything if they don't successfully install.

Also added a missing top-level dependency on `web3`.

Looks like the biggest overhead, by far, was Node 10's IPC data serialization. Simply by running on node 12, the walltime performance overhead between running in-process and running via worker are far closer.
